loader: return package format from defaultResolve if known

dygabo opened the pull request:
This is a proposed modification of defaultResolve to return the package
format in case it has been found during package resolution.
The format will be returned as described in the documentation:
https://nodejs.org/api/esm.html#resolvespecifier-context-defaultresolve
There is one new unit test as well:
test/es-module/test-esm-resolve-type.js

Please consider this PR as a **discussion base** if this feature is needed.
The approach was to only return the format from defaultResolve if 
a package.json file has been found and parsed.
All other cases only return url as before.

All unit tests are green, the new test passes as well.

The modification for recomputing the format type found in 
`lib/internal/modules/esm/load.js` is a temporary solution for the 
`make test` because highlight.js seems to be a dual package but 
declares `type: commonjs` in its package.json:
https://github.com/highlightjs/highlight.js/blob/main/package.json#L35

aduh95 commented:
@nodejs/loaders

GeoffreyBooth commented:
> The modification for recomputing the format type found in `lib/internal/modules/esm/load.js` is a temporary solution for the `make test` because highlight.js seems to be a dual package but declares `type: commonjs` in its package.json

If the goal is to return the format “if known,” this edge case strikes me as a clue that perhaps the logic here is wrong or incomplete; that maybe the format isn’t actually known for certain conditions.

dygabo commented:
> > The modification for recomputing the format type found in `lib/internal/modules/esm/load.js` is a temporary solution for the `make test` because highlight.js seems to be a dual package but declares `type: commonjs` in its package.json
> 
> If the goal is to return the format “if known,” this edge case strikes me as a clue that perhaps the logic here is wrong or incomplete; that maybe the format isn’t actually known for certain conditions.

> 

my interpretation here was that maybe the package.json of highlight.js should not define the type in its `package.json` since it is a dual package and commonjs is the default type in node anyway. as I am rather new to node I don't know if this conclusion is correct. 

is that scenario correct from the package implementation side? removing the `"type": "commonjs",` line from `package.json` of `highlight.js` solves it.

for the case:
main package.json contains:
```json
  "type": "commonjs",
[...]
  "exports": {
    ".": {
      "require": "./lib/index.js",
      "import": "./es/index.js"
    },
```
and `es/package.json` contains:
```
"type": "module"
```

is in this case the correct result "module" in case of import? If that is the case it might be that there is a bug in defaultResolve for this case even without the changes in this PR since it solves to "commonjs" (I could look into it if needed). that is not sent out and ` load` recomputes based on path and gets to the correct "module" result.

please let me know how to proceed.

dygabo commented:
I looked a bit more into this and I think I found the part of the code where this happens:
- `getPackageConfig` creates an entry in  `packageJSONCache` including `main` and `type` that are read from the main `package.json` file (in the specific testcase: `./lib/index.js` and `commonjs` respectively). 
- since the `package.json` contains an "export" section, `packageExportsResolve`  gets called and this solves correctly the main script to be in the specific case `./es/index.js` but it does not update the package type in the cache with the value in the `./es/package.json`. In fact the code in `packageExportsResolve` does not look for the existence of a "deeper" package.json

I attach here a screenshot of the callstack where the script resolution happens because I think it explains more than I can do in text form:
![image](https://user-images.githubusercontent.com/94833492/143837094-4b4b48d4-d2b8-4358-88d7-98469a5700d6.png)

if I find a feasible solution I will try to make a proposal, otherwise kindly waiting for input on this.

aduh95 commented:
> my interpretation here was that maybe the package.json of highlight.js should not define the type in its `package.json` since it is a dual package and commonjs is the default type in node anyway. as I am rather new to node I don't know if this conclusion is correct.

The `"type"` field in the `package.json` is useful only to determine how to parse `.js` file. When loading a  `.js` file, node has to look for the "closest"[^1] `package.json` to know if it has a `"type"` set to `"module"`, otherwise it's parsed as CJS.
[^1]: the one in the same directory, or one of the parent directories, which may or may not be the same `package.json` that defines the `"exports"`.

Node.js docs recommends to always set the `"type"` field, regardless if the package is dual or CJS-only: https://github.com/nodejs/node/blob/e31d1cb55da6ad9d05ed98d22164c01501ca2b39/doc/api/packages.md#L80-L84

dygabo commented:
> The `"type"` field in the `package.json` is useful only to determine how to parse `.js` file. When loading a `.js` file, node has to look for the "closest"[1](#user-content-fn-1-76e64c98206c83b6df7b0ea014b07f69) `package.json` to know if it has a `"type"` set to `"module"`, otherwise it's parsed as CJS.
> 
> Node.js docs recommends to always set the `"type"` field, regardless if the package is dual or CJS-only:
> 
> https://github.com/nodejs/node/blob/e31d1cb55da6ad9d05ed98d22164c01501ca2b39/doc/api/packages.md#L80-L84
> 
> ## Footnotes
> 1. the one in the same directory, or one of the parent directories, which may or may not be the same `package.json` that defines the `"exports"`. [↩](#user-content-fnref-1-76e64c98206c83b6df7b0ea014b07f69)

Thanks for the tips. I updated the PR with one commit considering your input. I also added an additional testcase to `test/es-module/test-esm-resolve-type.js` to reflect the situation with the dual package and the additional `package.json` file. 

All the tests are green and I look forward to hearing your thoughts on this.

bmeck commented on test/fixtures/es-modules/package-without-pjson/index.js:
```diff
@@ -0,0 +1,3 @@
+const identifier = 'package-without-pjson';
```
this needs to ensure not package.json exists above it https://github.com/nodejs/node/blob/fc4a7927288048089eaa6bf55177d7b4a3463609/test/common/index.js#L738

dygabo replied:
done with https://github.com/nodejs/node/pull/40980/commits/59350655e1d29fee1c56e6c0e194f28aed49b88d

bmeck commented on test/es-module/test-esm-resolve-type.js:
```diff
@@ -0,0 +1,146 @@
+            moduleType,
+            expectedResolvedType ] = testVariant;
+    tmpdir.refresh();
+    const rel = (file) => path.join(process.cwd(), file);
```
this should be using the tmpdir and not cwd

dygabo replied:
done with https://github.com/nodejs/node/pull/40980/commits/59350655e1d29fee1c56e6c0e194f28aed49b88d

bmeck commented on test/es-module/test-esm-resolve-type.js:
```diff
@@ -0,0 +1,146 @@
+  [ '/es-modules/package-type-commonjs/index.js', undefined ],
+  [ '/es-modules/package-without-type/index.js', undefined ],
+  [ '/es-modules/package-without-pjson/index.js', undefined ],
+].forEach((testVariant) => {
```
I'm a bit confused as to why these should not return the package.json "type" based format of the url in question.

dygabo replied:
that's just because this is the current behavior of defaultResolve and this test section only ensures that this is not changed by the modifications here. Resolving by fully qualified path does not currently need to look for `package.json` and implicitly does not know the `"type"` value. The goal was not to change the current resolution behavior, just to expose the type if it was found.

bmeck reviewed (approved):
Seems fine to me pending, a comment in the test case explaining what is going on regarding 2 package.json files might help though since it is rather hard to read.

dygabo commented:
> Seems fine to me pending, a comment in the test case explaining what is going on regarding 2 package.json files might help though since it is rather hard to read.

I added now a comment to the source file explaining the test scenario and also check the returned url value: https://github.com/nodejs/node/pull/40980/commits/022a2eb0990762096b38fb87b05dbd0fdb368bf3

aduh95 commented:
I'm not sure I understand this change, it looks like the added `.format` property is not used anywhere, so is it actually useful? Why are we adding it?

dygabo commented:
> I'm not sure I understand this change, it looks like the added `.format` property is not used anywhere, so is it actually useful? Why are we adding it?

it actually has an effect on the behaviour of `defaultLoad` but more than that it might be of advantage for external loaders added with `--loader` to have this information in their implementation after calling the defaultResolve in the resolve hook.

Flarna commented on test/es-module/test-esm-resolve-type.js:
```diff
@@ -0,0 +1,192 @@
+  assert.ok(resolveResult.url.includes('my-dual-package/es/index.js'));
+
+  // cleanup
+  fs.unlinkSync(esScript);
```
Could we move the cleanup also into the finally block to avoid leftovers in case of a failed test.
Maybe deleting the whole `tmpdir` recursive would be easier?

dygabo replied:
that is a good point, makes it cleaner. I pushed the change in https://github.com/nodejs/node/pull/40980/commits/33a457247c1d7ddb8b59d9ce435b99c2353cf270

aduh95 commented:
> > I'm not sure I understand this change, it looks like the added `.format` property is not used anywhere, so is it actually useful? Why are we adding it?
> 
> it actually has an effect on the behaviour of `defaultLoad` but more than that it might be of advantage for external loaders added with `--loader` to have this information in their implementation after calling the defaultResolve in the resolve hook.

Could you add a test that doesn't use `--expose-internals` that illustrate that? I think it would help me understand.

JakobJingleheimer commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -463,6 +463,21 @@ const patternRegEx = /\*/g;
+    try {
+      format = getPackageType(resolved);
+    } catch (err) {
+      if (err.code === 'ERR_INVALID_FILE_URL_PATH') {
+        throw new ERR_INVALID_MODULE_SPECIFIER(
+          resolved, 'must not include encoded "/" or "\\" characters', base);
+      }
+      throw err;
```
why create a new error?

dygabo replied:
that is to satisfy the assertion in the existing testcases (and avoid any undesired side-effects): https://github.com/nodejs/node/blob/8ac9aef8b1a323399ed1d67fb3a90c6c4d9e4d7a/test/es-module/test-esm-imports.mjs#L69

JakobJingleheimer commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -1031,7 +1064,10 @@ function defaultResolve(specifier, context = {}, defaultResolveUnused) {
-  return { url: `${url}` };
+  return {
+    url: `${url}`,
+    ...(format !== undefined && format !== null) && { format: format }
```
@aduh95 this is where format would be exposed to userland loaders

```js
async function resolve(specifier, context, defaultResolve) {
    const {
        format, // courtesy of this PR
        url,
    } = defaultResolve(specifier, context);
}
```

JakobJingleheimer reviewed (approved):
Cool, thanks for this!

Custom resolve hooks can already return an optional `format` property, so this makes sense to me. I think there wasn't a particular reason we did not already update `defaultResolve()` to do this—I think maybe we were planning to do in a follow-up (so the original PR, that was already getting quite large, could land) and @dygabo got to it first 🙂

Approach and goal look good to me. Just a few hygiene nits.

aduh95 commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -907,12 +936,15 @@ function moduleResolve(specifier, base, conditions, preserveSymlinks) {
   }
   if (resolved.protocol !== 'file:')
-    return resolved;
-  return finalizeResolution(resolved, base, preserveSymlinks);
+    return { url: resolved };
+  return {
+    url: finalizeResolution(resolved, base, preserveSymlinks),
+    ...(format !== undefined && format !== null) && { format: format }
+  };
```
I think we should add brackets, I'm not sure the suggestion above really improves readability, having brackets should remove any ambiguity.

JakobJingleheimer replied:
Sure, or that. Just the indented version with nothing was my concern 🙂

dygabo replied:
I agree with the readability argument and also it is IMO good practice to have curly braces also on one-statement if blocks. This was like this in existing code and there is a number of places where it is following the same pattern. Did not want to start reformatting too much here. But since we are at this point in code I think it is good to fix it here.

Flarna commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -631,12 +650,12 @@ function packageExportsResolve(
       packageJSONUrl, target, '', packageSubpath, base, false, false, conditions
     );
-    if (resolved === null || resolved === undefined)
+    if (resolveResult === null || resolveResult === undefined)
```
```suggestion
    if (resolveResult == null)
```

Flarna commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -717,11 +737,11 @@ function packageImportsResolve(name, base, conditions) {
         );
-        if (resolved !== null && resolved !== undefined)
-          return resolved;
+        if (resolveResult !== null && resolveResult !== undefined)
```
```suggestion
        if (resolveResult != null)
```

Flarna commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -747,11 +767,12 @@ function packageImportsResolve(name, base, conditions) {
+                                                     bestMatchSubpath,
+                                                     bestMatch, base, true,
+                                                     true, conditions);
+          if (resolveResult !== null && resolveResult !== undefined)
```
```suggestion
          if (resolveResult != null)
```

Flarna commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -907,12 +936,15 @@ function moduleResolve(specifier, base, conditions, preserveSymlinks) {
+    return { url: resolved };
+  return {
+    url: finalizeResolution(resolved, base, preserveSymlinks),
+    ...(format !== undefined && format !== null) && { format: format }
```
```suggestion
    ...(format != null) && { format: format }
```

dygabo commented:
> > > I'm not sure I understand this change, it looks like the added `.format` property is not used anywhere, so is it actually useful? Why are we adding it?
> > 
> > 
> > it actually has an effect on the behaviour of `defaultLoad` but more than that it might be of advantage for external loaders added with `--loader` to have this information in their implementation after calling the defaultResolve in the resolve hook.
> 
> Could you add a test that doesn't use `--expose-internals` that illustrate that? I think it would help me understand.

I cannot make a unit-test without the expose-internals in this case but I think it is explained by @JakobJingleheimer :
https://github.com/nodejs/node/pull/40980#pullrequestreview-823302229

please let me know if that clarifies the use-case

aduh95 commented:
> I cannot make a unit-test without the expose-internals in this case

Sure you can. For example, you could check that `format` and `resolved` are returned here: https://github.com/nodejs/node/blob/2cc7a91a5d855b4ff78f21f1bb8d4e55131d0615/test/fixtures/es-module-loaders/hooks-custom.mjs#L69

dygabo commented:
review comments solved with: https://github.com/nodejs/node/pull/40980/commits/3894a731567d08961ea6eb9a7000f7dea54afc46

dygabo commented:
> > I cannot make a unit-test without the expose-internals in this case
> 
> Sure you can. For example, you could check that `format` and `resolved` are returned here:
> 
> https://github.com/nodejs/node/blob/2cc7a91a5d855b4ff78f21f1bb8d4e55131d0615/test/fixtures/es-module-loaders/hooks-custom.mjs#L69

thank you for the hint. If this would serve the purpose of you understanding the modification I would try to squeeze in some more tome for this next week. I thought the comment of @JakobJingleheimer explained it quite well and your question would be clarified by that. 

Apart from that since I also am kind of on a tight schedule I would skip it as IMO the automated test with `--experimental-loader` does not bring additional value. That's because the existing testcase covers the modification as well. 

Please all let me know your thoughts on this. If the additional testcase is needed I will find an evening in the upcoming week for that.

GeoffreyBooth commented:
If you don’t mind, I’d also like a test that includes a user loader. This functionality is very closely related to user loaders and we should include coverage for it, because I think it would be important on the loaders side that any new loaders refactors are aware of this property and our loaders API changes don’t break any user loaders that rely on it.

I’ll try to find time to review this PR soon.

aduh95 commented:
Benchmark CI: https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/1060/

Flarna commented:
@dygabo Seems the added test fails on some systems:
https://ci.nodejs.org/job/node-test-commit-aix/39198/nodes=aix72-ppc64/testReport/junit/(root)/test/es_module_test_esm_resolve_type/
https://ci.nodejs.org/job/node-test-binary-windows-js-suites/12618/RUN_SUBSET=1,nodes=win10-COMPILED_BY-vs2019/testReport/junit/(root)/test/es_module_test_esm_resolve_type/

dygabo commented:
@GeoffreyBooth 
> If you don’t mind, I’d also like a test that includes a user loader.

I will try to prepare something for that and push it here.

> One question: what effect, if any, does this PR have on performance?

this will have only very limited impact on performance (or none). It should be slightly better as in some circumstances `load` does not need to redo some work that has been done by `resolve` already but I don't think that has relevant impact. The results from the benchmark CI triggered by @aduh95 confirm this.

@Flarna :
> Seems the added test fails on some systems

I will have a look and post here the findings.

dygabo commented:
I pushed a fix for the problem that caused the new test to fail on some platforms: https://github.com/nodejs/node/pull/40980/commits/5a87b59dab3f74a08475f24311baba09e13ac667. Can someone please restart the verification build (I could test it in my environment on windows 10 and ubuntu).

Please rerun build to verify if it works on all platforms correctly. Thanks

Flarna commented:
@dygabo There is still a relevant fail in your test because `process.chdir` is not supported in worker threads: https://ci.nodejs.org/job/node-test-commit-custom-suites-freestyle/23289/testReport/junit/(root)/test/es_module_test_esm_resolve_type/

If you can't get rid of `process.chdir` in this test it's most likely best to just skip it in workers like [here](https://github.com/nodejs/node/blob/32f7218388216b650bd27e0260401c60f41f73fd/test/parallel/test-fs-realpath.js#L27-L28).

dygabo commented:
thanks @Flarna for looking that up in the logs and for the solution hint. I did not know that check for all tests running on workers exist in the project.
Is there a possibility for me to try all those test scenarios somehow offline and only push once they're all green?

I suppose it should not be far from a stable state by now but it gets _embarrassing_ at times :) And would avoid the big loops for stuff that I would probably be able to fix before publishing

For this time around I disabled the test on workers according to your hint. See: https://github.com/nodejs/node/pull/40980/commits/075fc3098401c3d4e565a3a786984b7387976bb9

Please restart the verification build to check how far we get this time ;)

aduh95 commented:
> Is there a possibility for me to try all those test scenarios somehow offline and only push once they're all green?

`tools/test.py --worker test/es-module/`

aduh95 commented on lib/internal/modules/esm/resolve.js:
```diff
@@ -463,6 +463,21 @@ const patternRegEx = /\*/g;
+      format = getPackageType(resolved);
+    } catch (err) {
+      if (err.code === 'ERR_INVALID_FILE_URL_PATH') {
+        throw new ERR_INVALID_MODULE_SPECIFIER(
+          resolved, 'must not include encoded "/" or "\\" characters', base);
```
nit: not sure if we really care, but we could be passing the original error as cause.
```suggestion
        const er = new ERR_INVALID_MODULE_SPECIFIER(
          resolved, 'must not include encoded "/" or "\\" characters', base);
        er.cause = err;
        throw er;
```

aduh95 reviewed (changes requested):
Officially adding request for change to make sure this does not land until the non-internal test part is added or if the loaders team decide it's not needed.

aduh95 commented on test/es-module/test-esm-loader-resolve-type.mjs:
```diff
@@ -0,0 +1,42 @@
+allowGlobals(global.getModuleTypeStats);
+
+const basePath =
+  path.dirname(url.fileURLToPath(import.meta.url)) + '/node_modules';
```
Usage of `URL` instead of `node:path` makes more sense in ES module files imho.
```suggestion
  new URL('./node_modules/', import.meta.url);
```

aduh95 commented on test/es-module/test-esm-loader-resolve-type.mjs:
```diff
@@ -0,0 +1,42 @@
+const basePath =
+  path.dirname(url.fileURLToPath(import.meta.url)) + '/node_modules';
+
+const rel = (file) => path.join(basePath, file);
```
```suggestion
const rel = (file) => new URL(file, basePath);
```

aduh95 commented on test/es-module/test-esm-loader-resolve-type.mjs:
```diff
@@ -0,0 +1,42 @@
+createDir(basePath);
+createDir(moduleDir);
+fs.cpSync(
+  `./test/fixtures/es-modules/${moduleName}`,
```
You need to `import * as fixtures from '../common/fixtures.mjs'` 
```suggestion
  fixtures.path(`es-modules/${moduleName}`),
```

aduh95 commented on test/fixtures/es-module-loaders/hook-resolve-type.mjs:
```diff
@@ -0,0 +1,19 @@
+    importedCJS++;
+  } 
+
+  return {url, ...(format != null) && {format} };
```
Can't we simplify that?
```suggestion
  return { url, format };
```

dygabo replied:
even if it only is a subtle difference for the case where format is not resolved by the defaultResolver I would still let it as it is. 
On the return object { url, format } for example hasOwnProperty('format') will answer true even if format is not returned by the defaultresolver. with the conditional as it currently is, hasOwnProperty('format') will return false hence keeping the behaviour of the default.

aduh95 replied:
I think it's a valid argument for code in `lib/`, but in this test it will never matter because we don't call `Object.hasOwn` on the result. I would still remove it here.

dygabo replied:
writing (_slightly_) less correct code just because it looks easier in syntax is imho not something we as a community should strive to achieve.
Besides, writing a custom loader is not part of the "hello world" Node.js example so it can be expected that people looking at these tests to find an example will understand that syntax immediatelly. I added a short _green text_ to explain why we chose the conditional spread syntax there

aduh95 replied:
How is it less correct to pass `format: undefined`?

dygabo replied:
One of the goals of the additional loader is to wrap the default loader and produce no observable side-effects. The fact that `hasOwnProperty('format')` returns different results if applied on the returned value of custom loader `resolve` and on original default `resolve` *is* an observable side-effect of introducing the additional loader. The fact that this is currently not used does not exclude the possibility that `lib\internal\modules\esm\loader.js` (or its replacement) will use it in the future.
I know it is theoretical but it does not make it less true. In fact I think we are overdiscussing it as it will not be of too much importance since we're talking of an idle sample code. Nonetheless, it is as I see it an argument for production-level code.
So if the condition to get the PR through is to return `{ url, format }` in the sample loader just let me know and I will do that commit.

aduh95 replied:
To be clear, this is not a blocking concern from me (this PR is only blocked by https://github.com/nodejs/node/pull/41132 as far as I'm concerned); I'm discussing this because I would be interested in writing loaders myself, to understand how to write them correctly.

> The fact that `hasOwnProperty('format')` returns different results if applied on the returned value of custom loader `resolve` and on original default `resolve` _is_ an observable side-effect of introducing the additional loader.


It would probably be simpler to return the original object then.

Alternative proposal: make the default loader always return an object with a `format` property, so `Object.hasOwn('format')` is always `true`. It's not uncommon is ECMAScript to return an object with a property whose value is `undefined`, e.g.:
```js
const obj = {};
Object.defineProperty(obj, 'property', { get() { return 4 } });
console.log(Object.hasOwn(Reflect.getOwnPropertyDescriptor(obj, 'property'), 'set')); // true
```

What do you think?

dygabo replied:
As a matter of fact my initial take on this would have been to always try to resolve the type based on the `package.json`. That means we have the defaultResolver *always* return the `format` as good as it can tell. But that breaks existing tests, it possibly creates other side-effects. Actually that was also a review comment from @bmeck on this PR: https://github.com/nodejs/node/pull/40980#discussion_r758607561 
Because this would be my first contribution to the project I decided to take it one step at a time. Only where currently existing code has this information, pass it out of the function. This keeps existing tests green and allows external loaders to benefit of the information computed by `defaultResolve` in some cases. This is a minimally invasive change.
If I were to choose, I would try to close this up one way or the other. If it gets merged, next steps can be discussed to go in the direction of always returning `format` from `defaultResolve`. But that comes with an interface change, documentation update, and so on. This also may have some impact on ongoing loaders group work as far as I can tell from where I stay. So I'm not sure that my guess is enough of an "educated guess" and for sure I'm not the one to make this decision. If the whole loaders mechanism gets reimplemented from scratch based on the chaining redesign and considered obsolete, it might be that it's just water down the drain.
After https://github.com/nodejs/node/pull/41132 gets merged, I can rebase and squash, remove the `'url'` import and we take it from there. Would that make sense?

aduh95 replied:
OK that makes sense. Would you agree https://github.com/nodejs/node/pull/40980#discussion_r766752811 simplify the code?

> After #41132 gets merged, I can rebase and squash, remove the `'url'` import and we take it from there. Would that make sense?

Yes that would be perfect (FYI you don't have to squash yourself, but if you prefer to do it that's fine by me), then we can run the CI and land this finally :)

dygabo replied:
> Would you agree [#40980 (comment)](https://github.com/nodejs/node/pull/40980#discussion_r766752811) simplify the code?

Yes, I will take this over with the next commit. Will do it with the rebase after your `fs` bugfix gets merged.

dygabo commented:
other review findings have been taken over. Please restart verification build on https://github.com/nodejs/node/pull/40980/commits/085733777008da90bb670e45d24e7c3b0b6feb21

aduh95 commented on test/es-module/test-esm-loader-resolve-type.mjs:
```diff
@@ -0,0 +1,42 @@
+        importedCJS: importedCJSBefore } = global.getModuleTypeStats();
+
+import(`${moduleName}`).finally(() => {
+  fs.rmSync(fileURLToPath(basePath), { recursive: true, force: true });
```
That is not necessary, all `fs` functions accept URLs as first parameter.
```suggestion
  fs.rmSync(basePath, { recursive: true, force: true });
```

aduh95 replied:
https://github.com/nodejs/node/pull/41132 has landed.

dygabo replied:
dependency resolved with commit: https://github.com/nodejs/node/pull/40980/commits/2d2c4754322117dd5ccf362206e45be42c0edfb8

aduh95 commented on test/es-module/test-esm-loader-resolve-type.mjs:
```diff
@@ -0,0 +1,42 @@
+import * as fixtures from '../common/fixtures.mjs';
+import { strict as assert } from 'assert';
+import * as fs from 'fs';
+import { fileURLToPath, URL } from 'url';
```
That's not necessary, `URL` is available on the global scope.
```suggestion
```

dygabo replied:
removed URL from that import, `fileURLToPath` is still needed because of the exception with `rmSync(URL)`

aduh95 commented on test/fixtures/es-module-loaders/hook-resolve-type.mjs:
```diff
@@ -0,0 +1,25 @@
+}
+
+export function resolve(specifier, context, next) {
+  const { url, format } = next(specifier, context);
+
+  if ((format != null && format == 'module') || specifier.endsWith('.mjs')) {
+    importedESM++;
+  } else if (format == null || format == 'commonjs') {
+    importedCJS++;
+  } 
+
+  /**
+   * the conditional spread syntax is preferred over `return { url, format }`
+   * because `format` is optional for the `resolve` hook. If the 
+   * `defaultResolve` call did not return a format this resolve will not return
+   * a `format` property on the return object either.
+   */
+  return { url, ...(format != null) && {format} };
```
```suggestion
  const nextResult = next(specifier, context);
  const { format } = nextResult;

  if (format === 'module' || specifier.endsWith('.mjs')) {
    importedESM++;
  } else if (format == null || format == 'commonjs') {
    importedCJS++;
  } 

  return nextResult;
```

guybedford commented:
I would really have appreciated a ping for review on this PR - please make sure to ping @nodejs/modules as well in future.

JakobJingleheimer commented:
Sorry @guybedford. I saw it was tagged `esm` so I thought it pinged you automatically (I get pinged for `loaders`).

danielleadams commented:
@dygabo this doesn't land cleanly in 16.x - do you mind creating a backport PR?

dygabo commented:
@danielleadams sure, I will look into it

dygabo commented:
Backport PR for `v16.x-staging`: #41752
